File: Leetcode/MaximumScoreAfterSplittingaString.rs
// https://leetcode.com/problems/maximum-score-after-splitting-a-string

struct Solution {}

impl Solution {
    pub fn max_score(s: String) -> i32 {
        s.chars()
            .fold(
                (0, 0, s.chars().filter(|c| c == &'1').count() as i32, 1),
                |(mut mx, mut zeros, mut ones, left_len), c| {
                    zeros += (c == '0') as i32;
                    ones -= (c == '1') as i32;
                    if left_len != s.len() {
                        mx = mx.max(zeros + ones);
                    }
                    (mx, zeros, ones, left_len + 1)
                },
            )
            .0
    }
}

File: Leetcode/MaximumScoreAfterSplittingaString.cpp
// https://leetcode.com/problems/maximum-score-after-splitting-a-string

class Solution {
   public:
    int maxScore(string s) {
        int ones = 0;
        for (char c : s) {
            if (c == '1') {
                ones += 1;
            }
        }

        int zeros = s[0] == '0';
        ones -= s[0] == '1';
        int ans = zeros + ones;
        for (int i = 1; i < s.size() - 1; ++i) {
            if (s[i] == '0') {
                zeros += 1;
            } else {
                ones -= 1;
            }
            ans = max(ans, zeros + ones);
        }

        return ans;
    }
};